# ICARUS
> Study of U3-1402 and DS-1062a Antibody Drug Conjugates in Advanced Breast and Lung Cancer Patients
[![License](https://img.shields.io/badge/License-BSD_3--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

ICARUS BREAST01 [NCT04965766](https://clinicaltrials.gov/ct2/show/NCT04965766) and ICARUS LUNG01
[NCT04940325](https://clinicaltrials.gov/ct2/show/NCT04940325) are two multicentric phase II French clinical trials
resulting from the collaboration between Gustave Roussy and Daiichi Sankyo.
The BREAST01 trial aims at assessing the activity of *patritumab deruxtecan* (U3-1402; targets HER3) in advanced breast
cancers, while the LUNG01 trials investigates *datopotamab deruxtecan* (DS-1062a; targets TROP2)
in advanced or non-resectable non-small cell lung cancers (NSCLC).

**Principal investigators**
- Dr. Barbara PISTILLI
- Pr. David PLANCHARD
- Dr. Guillaume MONTAGNAC

## Repository organisation

### data

- `cln` all patient- and sample-related raw and curated tables
- `histopathology` the raw slides, annotations, and tables necessary for the histopathological analysis
- `resources` various resources files such as the capture BED for the WES, lists of genes of interests, list of
  oncogenic pathways etc.
- `rna` all analysis-ready tables generated by the RNAseq pipeline
- `wes` all analysis-ready tables generated by the WES pipeline and excel summary tables of QC per each batch executed.

See `data/README.md` to learn about data access.


### code

Folder hosting the bioinformatic pipelines employed on the sequencing files and the analysis scripts developed for
the project.

- `common` folder with code for external tools added as git sub modules and with scripts for defining utils
- `histopathological_analysis` a set of scripts to perform all the analysis related to the histopathological data. See
  `code/histopathological_analysis/README.md` for more details.
- `pipeline_rna` a Snakemake pipeline performing a comprehensive analysis of bulk RNA-seq data in order to quantify gene
  expression using SALMON+Tximport before proceeding to DEA (DESeq2), GSEA, and immune deconvolution.
- `pipeline_wes` a Snakemake pipeline performing a comprehensive analysis of WES data in order to identify, filter, and
  annotate somatic mutations, copy-number alterations, purity and ploidy in each pair of tumor/normal. It may also run
  in tumor-only mode for most all analyses. See `pipelines/wes/README.md` for more details.
- `stats_paper` folder with a single script which computes various statistics presented in the main text or the methods.
- `tables_paper` folder with scripts for generating tables supporting the paper (supplementary tables) or supporting
  data releases (EGA, cbioportal, etc).
- `wes_analysis` folder with scripts for drawing various types of oncoplots.

# Contributors

- Bastien JOB <https://github.com/aoumess>; bastien.job@gustaveroussy.fr
- Marc DELOGER <https://github.com/mdeloger>; marc.deloger@gustaveroussy.fr
- Yoann PRADAT <https://github.com/ypradat>; yoann.pradat@gustaveroussy.fr
- Lo√Øc Le Bescond <https://github.com/loic-lb>; loic.le-bescond@gustaveroussy.fr
